File: src/lib/ai/keystore.ts
/**
 * Where a credential lives, per platform.
 *
 * Desktop has the OS keychain — encrypted at rest, scoped to the user account.
 * The browser has nothing better than `localStorage`, and the UI says so rather
 * than implying a safety that is not there.
 *
 * See docs/adr/0010-secrets-handling.md.
 */

import { isTauri } from "@/lib/utils";
import { Secret } from "@/lib/secrets";

export type SecretKey =
  | "anthropicApiKey"
  | "openaiApiKey"
  | "groqApiKey"
  | "openrouterApiKey"
  | "googleApiKey"
  | "customApiKey";

/** Where credentials are being kept right now. Shown to the user. */
export type KeystoreKind = "keychain" | "browser";

export function keystoreKind(): KeystoreKind {
  return isTauri() ? "keychain" : "browser";
}

/** Prefix for the browser fallback, kept distinct from the settings blob. */
const WEB_PREFIX = "lexicon.secret.";

export async function getSecret(key: SecretKey): Promise<Secret | null> {
  if (isTauri()) {
    try {
      const { invoke } = await import("@tauri-apps/api/core");
      const value = await invoke<string | null>("get_secret", { key });
      return value ? new Secret(value) : null;
    } catch (error) {
      // A keychain that refuses to open is worth surfacing in the console, but
      // must not stop the app: the user can still use a local provider.
      console.warn(`keychain read failed for ${key}`, error);
      return null;
    }
  }
  const value = readWeb(key);
  return value ? new Secret(value) : null;
}

export async function setSecret(key: SecretKey, value: string): Promise<void> {
  if (isTauri()) {
    const { invoke } = await import("@tauri-apps/api/core");
    await invoke("set_secret", { key, value });
    return;
  }
  if (value) {
    localStorage.setItem(WEB_PREFIX + key, value);
  } else {
    localStorage.removeItem(WEB_PREFIX + key);
  }
}

export async function deleteSecret(key: SecretKey): Promise<void> {
  await setSecret(key, "");
}

/**
 * Move a credential out of the old settings blob.
 *
 * Before ADR 0010 the Anthropic key lived in `lexicon.settings.v1`. Leaving a
 * copy there would mean the keychain was an addition rather than a move — and
 * the old copy is the one that leaked into backups.
 *
 * Returns true if something was migrated, so the caller can say so.
 */
export async function migrateLegacySecrets(): Promise<boolean> {
  let migrated = false;
  try {
    const raw = localStorage.getItem("lexicon.settings.v1");
    if (!raw) return false;
    const parsed = JSON.parse(raw) as { state?: Record<string, unknown> };
    const legacy = parsed.state?.anthropicApiKey;

    if (typeof legacy === "string" && legacy.trim()) {
      await setSecret("anthropicApiKey", legacy);
      delete parsed.state!.anthropicApiKey;
      localStorage.setItem("lexicon.settings.v1", JSON.stringify(parsed));
      migrated = true;
    }
  } catch (error) {
    // A failed migration must not brick startup. The key stays where it was,
    // which is no worse than before this ran.
    console.warn("could not migrate the stored API key", error);
  }
  return migrated;
}

function readWeb(key: SecretKey): string | null {
  try {
    return localStorage.getItem(WEB_PREFIX + key);
  } catch {
    return null;
  }
}

File: src-tauri/src/keystore.rs
//! Credentials in the OS keychain rather than `localStorage`.
//!
//! On desktop there is somewhere genuinely better than a JSON blob on disk:
//! the platform credential store, which is encrypted at rest and scoped to the
//! user account. The browser has no equivalent, and the app says so there
//! rather than implying a safety it does not have.
//!
//! See docs/adr/0010-secrets-handling.md.

use keyring::Entry;

/// One service name for the whole app, so a user can find and revoke these.
const SERVICE: &str = "com.lexicon.app";

/// Credential names this app may touch.
///
/// An allowlist, so a compromised frontend cannot turn these commands into a
/// way to read arbitrary entries out of the user's keychain — which would be a
/// far worse hole than the one this feature closes.
const ALLOWED_KEYS: &[&str] = &[
    "anthropicApiKey",
    "openaiApiKey",
    "groqApiKey",
    "openrouterApiKey",
    "googleApiKey",
    "customApiKey",
];

fn entry(key: &str) -> Result<Entry, String> {
    if !ALLOWED_KEYS.contains(&key) {
        return Err(format!("{key} is not a credential this app stores"));
    }
    Entry::new(SERVICE, key).map_err(|e| format!("keychain unavailable: {e}"))
}

/// Store a credential. An empty value deletes it.
#[tauri::command]
pub fn set_secret(key: String, value: String) -> Result<(), String> {
    let entry = entry(&key)?;
    if value.is_empty() {
        // Deleting an absent entry is success, not failure.
        return match entry.delete_credential() {
            Ok(()) => Ok(()),
            Err(keyring::Error::NoEntry) => Ok(()),
            Err(e) => Err(format!("could not clear {key}: {e}")),
        };
    }
    entry
        .set_password(&value)
        .map_err(|e| format!("could not save {key}: {e}"))
}

/// Read a credential, or `None` if it was never stored.
#[tauri::command]
pub fn get_secret(key: String) -> Result<Option<String>, String> {
    match entry(&key)?.get_password() {
        Ok(value) => Ok(Some(value)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(format!("could not read {key}: {e}")),
    }
}

/// Remove a credential.
#[tauri::command]
pub fn delete_secret(key: String) -> Result<(), String> {
    match entry(&key)?.delete_credential() {
        Ok(()) => Ok(()),
        Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(format!("could not delete {key}: {e}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_known_credentials_are_addressable() {
        assert!(entry("anthropicApiKey").is_ok());
        // The allowlist is the security boundary: without it these commands
        // would read any entry in the user's keychain.
        for key in [
            "",
            "ssh-key",
            "../../other-service",
            "GitHub",
            "anthropicApiKey ",
        ] {
            assert!(entry(key).is_err(), "{key} should be rejected");
        }
    }

    #[test]
    fn every_allowed_key_is_a_credential_name() {
        assert!(ALLOWED_KEYS.iter().all(|k| k.ends_with("ApiKey")));
    }
}
